src/arlo_e2e/ray_map_reduce.py

```diff
@@ -0,0 +1,341 @@
+                ]
+                running_jobs = new_job_refs + running_jobs
+
+            # Next, we'll wait for half a second and see how we're doing.
+            tmp: Tuple[List[ObjectRef], List[ObjectRef]] = ray.wait(
+                running_jobs, num_returns=len(running_jobs), timeout=0.5
+            )
+            # TODO: Ray version 2 (dev builds) add an option here, fetch_local=False,
+            #  which seems like it would be relevant to us. It's not available in the
+            #  version 1.1 release. Unclear whether that fetching is happening in
+            #  the 1.1 release of Ray or not. Hopefully not.
+
+            new_ready_refs, pending_refs = tmp
+            ready_refs = ready_refs + new_ready_refs
+            num_ready_refs = len(ready_refs)
+            num_pending_refs = len(pending_refs)
+
+            if (
+                num_ready_refs == 1
+                and num_pending_refs == 0
+                and len(pending_map_shards) == 0
+            ):
+                # terminal case: we have one result and nothing pending; we're done!
+                result = ready_refs[0]
+                break
+
+            if num_ready_refs >= 2:
+                # general case: we have some results to reduce
+
+                reduce_shards = shard_list_uniform(ready_refs, self._reduce_shard_size)
+                size_one_shards = [s for s in reduce_shards if len(s) == 1]
+                usable_shards = [s for s in reduce_shards if len(s) > 1]
+                total_usable = sum(len(s) for s in usable_shards)
+
+                if progress_actor:
+                    progress_actor.update_total.remote(
+                        self._reduction_description, total_usable
+                    )
+
+                partial_reductions = [
+                    _ray_mr_reduce.remote(
+                        context_ref, progress_actor, reduction_desc_ref, *rs
+                    )
+                    for rs in usable_shards
+                ]
+
+                running_jobs = partial_reductions + pending_refs
+                ready_refs = [s[0] for s in size_one_shards]
+            else:
+                # Annoying case: we have exactly one ready result, one or more pending
+                # results, and nothing to reduce, so we'll just go back and ray.wait
+                # again for more results to show up.
+                running_jobs = pending_refs
+                pass
+
+        assert result is not None, "reducer fail: exited loop with no result!"
+
+        if progressbar:
+            progressbar.close()
+        result_local: R = ray.get(result)
+        return result_local
+
+    def __init__(
+        self,
+        context: MapReduceContext[T, R],
+        use_progressbar: bool = False,
+        input_description: str = DEFAULT_INPUT_STR,
+        reduction_description: str = DEFAULT_REDUCTION_STR,
+        max_tasks: int = 5000,
+        map_shard_size: int = 4,
+        reduce_shard_size: int = 10,
```

Comment:
it might be fun to play with automatic tuning of these. I'm sure they're a research paper out there discussing it.

Reply:
Yeah, I thought about setting a timer, calling `ray.workers()` every so often, and automatically adjusting this thing. At some point, though, this whole map-reduce structure is something that the Ray people need to support themselves, since they know way better than we do exactly how much excess of ready-but-not-running tasks you need to keep around to keep the pipelines full.